techlibs/gatemate/gatemate_bramopt.cc

```diff
@@ -0,0 +1,146 @@
+/*
+ *  yosys -- Yosys Open SYnthesis Suite
+ *
+ *  Copyright (C) 2021  Cologne Chip AG <support@colognechip.com>
+ *
+ *  Permission to use, copy, modify, and/or distribute this software for any
+ *  purpose with or without fee is hereby granted, provided that the above
+ *  copyright notice and this permission notice appear in all copies.
+ *
+ *  THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES
+ *  WITH REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF
+ *  MERCHANTABILITY AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR
+ *  ANY SPECIAL, DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
+ *  WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER IN AN
+ *  ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
+ *  OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
+ *
+ */
+
+#include "kernel/yosys.h"
+#include "kernel/sigtools.h"
+
+USING_YOSYS_NAMESPACE
+PRIVATE_NAMESPACE_BEGIN
+
+bool binary_pred(SigBit s1, SigBit s2)
+{
+	return s1 == s2;
+}
+
+static void proc_glwren(Module *module)
+{
+	std::vector<Cell*> ram_cells;
+
+	// Gather RAM cells
+	for (auto cell : module->cells())
+	{
+		if (cell->type.in(ID(CC_DPSRAM_20K), ID(CC_DPSRAM_40K))) {
+			ram_cells.push_back(cell);
+		}
+	}
+
+	// Convert bitmask signals
+	for (auto cell : ram_cells)
+	{
+		std::vector<SigBit> mska = cell->getPort(ID(A_BM)).to_sigbit_vector();
```

Comment:
this transformation is unsound — consider the following case:

```
reg [15:0] mem[0:1023];

initial $readmemh(...);

always @(posedge clk)
    if (we)
        mem[15:8] <= wdata;
```

In this case, the transformation as written would apply, overwriting all write enables with const-1, and making the memory write port overwrite the low 8 bits of the word on every write, which should never be changed

What *should* be done here, to get a sound transformation:

- check if there is exactly one SigBit other than S0 in the write enable, remember which indices are this SigBit and which are S0
- replace the main write enable with that SigBit, replace the corresponding bits within write masks with S1, but *don't* replace bits that were S0 originally
- if S0 is harder to materialize in P&R than S1 for some reason, an additional optimization could replace them with S1 for bit lanes which are never read (ie. their read port data bits are connected to nothing)

Reply:
With the latest changes in `brams_map.v`in 09dd91e, the `gatemate_bramopt` pass has become obsolete. The current result is optimal for our p&r. We would therefore like to tend to remove this pass, if you don't mind.

Reply:
SGTM